src/data/finsum.lean

```diff
@@ -0,0 +1,657 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kexing Ying, Kevin Buzzard
+-/
+
+import data.set.finite
+import algebra.big_operators
+import data.support
+import data.finsupp
+
+/-!
+# Finite sums over types and sets
+
+We define sums over types and subsets of types, with no finiteness hypotheses. All infinite
+sums are defined to be junk values (i.e. zero). This approach is sometimes easier to use
+than `finset.sum` when issues arise with `finset` and `fintype` being data.
+
+## Main definitions
+
+Variables:`
+
+`(α : Type u)`
+`(s : set α)`
+`(M : Type v) [add_comm_monoid M]`
+`(f : α → M)`
+
+Definitions in this file:
+
+* `univ' α : finset α` : `α` as a finset if `α` is finite, and the empty finset otherwise
+
+* `finsum f : M` : the sum of `f x` as `x` ranges over the support of `f`, if it's finite.
+   Zero otherwise.
+
+* `finsum_in s f : M` : the sum of `f x` as `x` ranges over the elements of `s` for which `f x ≠ 0`,
+  if this is a finite sum. Zero otherwise.
+
+## Notation
+
+* `∑ᶠ i, f i` and `∑ᶠ i : α, f i` for `finsum f`
+
+* `∑ᶠ i in s, f i` for `finsum_in s f`
+
+## Implementation notes
+
+There is no `finprod` right now. This is because `function.support` and `finsupp` etc
+are all focussed on 0, and there are no analogous definitions for 1.
+
+## Todo
+
+We did not add `is_finite (X : Type) : Prop`, because it is simply `nonempty (fintype X)`.
+There is work on `fincard` in the pipeline, which returns the cardinality of `X` if it
+is finite, and 0 otherwise.
+
+## Tags
+
+finsum, finsum_in, univ', finiteness
+-/
+
+open_locale classical
+
+universes u v w
+
+/-! # finset.univ' -- noncomputable univ (empty set if infinite) -/
+
+namespace finset
+
+/-- `univ' α` is the finset corresponding to all of α if α is finite, and the empty set otherwise.
+Note that it is noncomputable -/
+noncomputable def univ' (α : Type u) : finset α :=
+if h : nonempty (fintype α) then (classical.choice h).elems else ∅
+
+/-- `set.to_finset s` is the finset corresponding to `s` if `s` is finite, and the empty set
+    otherwise -/
+noncomputable def set.to_finset {α : Type u} (s : set α) : finset α :=
+if h : s.finite then h.to_finset else ∅
+
+variable {α : Type u}
+
+/-- `univ' α` is finset.univ if `α` is a fintype. -/
+lemma univ'_eq_univ [h : fintype α] : univ' α = univ :=
+by convert (dif_pos (nonempty.intro h))
+
+lemma univ'_eq_empty (h : ¬ nonempty (fintype α)) : univ' α = ∅ :=
+dif_neg h
+
+@[simp] lemma mem_univ' [h : fintype α] (x : α) : x ∈ univ' α :=
+(@univ'_eq_univ _ h).symm ▸ mem_univ _
+
+end finset
+
+/-! # noncomputable finite sums -/
+
+section finsum
+
+variables {α : Type u} {M : Type v} [add_comm_monoid M]
+
+/-- Sum of `f x` as `x` ranges over the elements of the support of `f`, if it's finite. Zero
+  otherwise. -/
+noncomputable def finsum (f : α → M) : M :=
+if h : (function.support f).finite then finset.sum (h.to_finset) f else 0
```

Comment:
Could this be expressed as `(function.support f).to_finset.sum f`, using the definition above it? It should still evaluate to 0 in the same circumstances.

Reply:
Apparently it can be!